---   
<img align="left" width="110"   src="https://upload.wikimedia.org/wikipedia/commons/c/c3/Python-logo-notext.svg"> 

<h1 align="center">Tools and Techniques for Data Science</h1>
<h1 align="center">Course: Python for Data Scientist</h1>

---

<h3><div align="right">Muhammad Sheraz(Data Scientist)</div></h3>    

<h1 align="center">Lecture 3.12 (Pandas-04)</h1>

<a href="https://colab.research.google.com/github/arifpucit/data-science/blob/master/Section-3-Python-for-Data-Scientists/Lec-3.12(Pandas-04-IO-with-CSV-EXCEL-and-JSON-Files).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## _IO with CSV EXCEL and JSON Files_

<img align="center" width="600" height="150"  src="images/fileformats.png" >

#### Read Pandas Documentation:
- General Info: https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html


- For `read_csv`: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html?highlight=read_csv#pandas.read_csv


- For `read_excel`: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_excel.html?highlight=read_excel#pandas.read_excel


- For `read_json`:https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.io.json.read_json.html?highlight=pandas%20read_json#pandas.io.json.read_json


- For `to_csv`: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_csv.html#pandas.DataFrame.to_csv



- For `to_excel`: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_excel.html?highlight=to_excel#pandas.DataFrame.to_excel


- For `to_json`: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_json.html?highlight=to_json

## Learning agenda of this notebook
[Pandas](https://pandas.pydata.org/) provides helper functions to read data from various file formats like CSV, EXCEL, JSON, HTML, SQL table, and many more.
1. Reading data from a CSV/TSV File
2. Reading a CSV file from a Remote System
3. Writing Contents of Dataframe to a CSV File
4. Reading data from an EXCEL File
5. Writing Contents of Dataframe to an EXCEL File
6. Reading data from a JSON File
7. Writing Contents of Dataframe to a JSON File

In [1]:
# To install this library in Jupyter notebook
import sys
!{sys.executable} -m pip install pandas --quiet

In [2]:
import pandas as pd
pd.__version__ , pd.__path__

('1.5.3', ['C:\\Users\\LENOVO\\anaconda3\\lib\\site-packages\\pandas'])

## 1. Reading from CSV/TSV Files
>**CSV**: A text file in which the values are separated by a comma or a tab character is called a CSV or a TSV file. Each line of the file is a data record and each record consists of one or more fields, separated by a specific character called separator. A CSV/TSV file is typically used to store tabular data (numbers and text), in which each line will have the same number of fields.

### a. Reading a  Simple CSV File
The `pd.read_csv()` method is used to read a comma-separated file into a DataFrame.
```
pd.read_csv(fname, delimiter=None, header='infer', skiprows=None , nrows=None , usecols=None,  footer='',...)

```

In [3]:
! cat datasets/classmarks.csv

'cat' is not recognized as an internal or external command,
operable program or batch file.


In [4]:
df.shape()

NameError: name 'df' is not defined

In [ ]:
#The `read_csv`, by default assumes that the file contains comma separated values, 
# and the first row of the file conatins names of columns, which will be taken as column labels
df = pd.read_csv('datasets/classmarks.csv')
df

**The `df.head(N)` method is used to select/display first `N` rows, based on `position`, i.e., the integer value corresponding to the position of the row (from 0 to n-1). The default value of `N` is 5.**

In [ ]:
df.head()

In [ ]:
df.head(3)

In [5]:
# For negative values of n, this method returns all rows except the last `n` rows, equivalent to df[:-n].
# The df has a total of 50 rows, so the following will return first 2 rows
df.head(-48)

NameError: name 'df' is not defined

**The `df.tail(N)` method is used to select/display last `N` rows, based on `position`, i.e., the integer value corresponding to the position of the row (from 0 to n-1). The default value of `N` is 5.**

In [6]:
# tail() method is useful for quickly verifying data, after sorting or appending rows.
df.tail()

NameError: name 'df' is not defined

In [7]:
df.tail(3)

NameError: name 'df' is not defined

In [ ]:
# For negative values of `n`, this function returns all rows except the first `n` rows, equivalent to df[n:]
# The df has a total of 50 rows, so the following will return last 3 rows
df.tail(-46)

### b.Reading a CSV File having a Delimter, other than Comma
- By default, the `read_csv()` expect comma as seperator. But if the CSV file has some other seperator or delimiter like (semi-collon or tab), it will raise an error.
- To handler the issue we need to pass specific value to the `delimiter` argument of `read_csv()` method.

In [ ]:
! cat datasets/classmarkswithtab.csv

In [ ]:
df = pd.read_csv('datasets/classmarkswithtab.csv')
df.head()

In [10]:
df = pd.read_csv('datasets/classmarkswithtab.csv', delimiter='\t', header=None)
df.head()

0       1        2     3     4        5     6
0  rollno  gender    group   age  math  english  urdu
1    MS01  female  group B  28.0  72.0       72  74.0
2    MS02  female  group C  33.0  69.0       90  88.0
3    MS03  female  group B  21.0   NaN       95  93.0
4    MS04    male  group A  44.0  47.0       57  44.0

### c. Reading a CSV File not having Column Labels
- By default the `read_csv()` method assume the first row of the file will contain column labels
- If this is not the case, i.e., the file do not contain column labels rather data, it will be dealt as column label
- Understand this in following example

In [ ]:
! cat datasets/classmarkswithoutcollabels.csv

In [ ]:
df = pd.read_csv('datasets/classmarkswithoutcollabels.csv')
df.head()

**To read such files, you have to pass the parameter `header=None` to the `read_csv()` method as shown below**

In [ ]:
df = pd.read_csv('datasets/classmarkswithoutcollabels.csv', header=None)
df.head()


**Now if you want to assign new column labels to make them more understandable, you can assign the list of column labels to the `columns` attribute of the dataframe object**

In [ ]:
col_names = ['rollno', 'gender', 'group', 'age', 'math', 'english', 'urdu']
df.columns = col_names
df.head()

### d. Reading a CSV File having Comments in the beginning
- You may get an error while reading a CSV file because someone may have added few comments on the top of the file. In pandas we can still read the data set by skipping few rows from the top.
- To deal with the ParseError, open the csv file in the text editor and check if you have some comments on the top.
- If yes, then count the number of rows to skip.
- While reading file, pass the parameter **skiprows = n** (number of rows in the beginninghaving comments to skip)
- While reading file, pass the parameter **skipfooter = n** (number of rows at the end having comments to skip)

In [ ]:
! cat datasets/classmarkswithtopcomments.csv

In [ ]:
# Try reading a csv file having 3 comments lines in the beginning.
df = pd.read_csv('datasets/classmarkswithtopcomments.csv')
df.head()

In [ ]:
# Try reading a csv file having 3 comments lines in the beginning.
df = pd.read_csv('datasets/classmarkswithtopcomments.csv', skiprows=3)
df.head()

### e. Reading a portion of CSV File in a Dataframe
- Suppose the dataset inside the csv file is too big and you don't want to spend that much time for reading that data
- Or might be your system crashes, when you try to load that much data
- Solution is read
    - Specific number of rows by passing `nrows` parameter to `read_csv()` method
    - Specific number of columns by passing `usecols` parameter to `read_csv()` method

In [ ]:
# Read just 10 rows from the csv file by passing the number of rows to read to `nrows` argument
df = pd.read_csv('datasets/classmarks.csv', nrows=10)
df.shape
#df.head()

In [ ]:
# Read specific columns from the csv file by passing a list of column names to `usecols` argument
df = pd.read_csv('datasets/classmarks.csv', usecols= ['rollno', 'group','english'])
df.shape
#df.head()

In [ ]:
# Ofcourse you can use both the parameters at the same time
df = pd.read_csv('datasets/classmarks.csv', nrows= 7, usecols= ['rollno', 'group','english'])
df.shape

In [ ]:
df

## 2. Reading a CSV File from a Remote System

In [ ]:
# To avoid URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED]..... 
import ssl
ssl._create_default_https_context = ssl._create_unverified_context

### a. Reading a CSV file from GitHub Gist

The `data1.csv` file actually resides on my GitHub Gist at following URL:
https://gist.githubusercontent.com/arifpucit/bbcb0bba0b5c245585b375f273f17876/raw/28ddebe991c86f7001178896329005ea174f2bde/data1.csv

Bitly is a URL shortening service that I have used to create a short link for easy usage in the cell below:

In [ ]:
import pandas as pd
myurl = 'https://bit.ly/31eYGTx'
df = pd.read_csv(myurl)
df

### b. Reading a CSV file from a Google Docs

- Google Sheet URL:  https://docs.google.com/spreadsheets/d/1H9ZTGVRXN3zuyP3cbJQnbX7JlqhqFNMOE_WwDP-PRTE/edit#gid=2084742287


In [ ]:
sheetID = '1H9ZTGVRXN3zuyP3cbJQnbX7JlqhqFNMOE_WwDP-PRTE'
sheetName = 'sheet1'
URL = 'https://docs.google.com/spreadsheets/d/{0}/gviz/tq?tqx=out:csv&sheet={1}'.format(sheetID, sheetName)

df = pd.read_csv(URL)
df.head()

In [ ]:
URL

## 3. Writing Contents of Dataframe to a CSV File
- The `pd.to_csv()` method is used to write the contents of a dataframe (with indices) to a CSV file.
- The only required argument is the file path.
- For details see help page or python documents (link given above)

In [ ]:
df_class = pd.read_csv('datasets/classmarks.csv')
df_class.head(7)

>- Let us create a new dataframe from above dataframe containing records of only group B

In [ ]:
mask = (df_class['group'] == 'group B')
mask.head(7)

In [ ]:
df_class_groupB = df_class.loc[mask]
df_class_groupB

In [ ]:
df_class_groupB.to_csv('datasets/classmarksgroupB.csv')

In [ ]:
df = pd.read_csv('datasets/classmarksgroupB.csv')
df

>To avoid writing the row indices column inside the file pass `index=False` argument to `to_csv()` method

In [ ]:
df_class_groupB.to_csv('datasets/classmarksgroupB.csv', index=False)

In [ ]:
df = pd.read_csv('datasets/classmarksgroupB.csv')
df

## 4. I/O with EXCEL Files
>**XLSX**: XLSX is a Microsoft Excel Open XML file format. It also comes under the Spreadsheet file format. It is an XML-based file format created by Microsoft Excel. In XLSX data is organized under the cells and columns in a sheet. Each XLSX file may contain one or more sheets. So a workbook can contain multiple sheets

In [ ]:
import sys
!{sys.executable} -m pip install xlrd xlwt openpyxl

### a. Reading a Simple Excel File

In [11]:
df = pd.read_excel(io='datasets/classmarks.xlsx')
df.head()

rollno  gender    group   age  math  english  urdu
0   MS01  female  group B  28.0  72.0       72  74.0
1   MS02  female  group C  33.0  69.0       90  88.0
2   MS03  female  group B  21.0   NaN       95  93.0
3   MS04    male  group A  44.0  47.0       57  44.0
4   MS05    male  group C  54.0  76.0       78   NaN

### b. Reading an Excel File having Comments in the beginning
- You may get an error while reading an Excel file because someone may have added few comments on the top of the file. In pandas we can still read the data set by skipping few rows from the top.
- To deal with the ParseError, open the Excel file in MS EXCEL and check if you have some comments on the top.
- If yes, then count the number of rows to skip.
- While reading file, pass the parameter **skiprows = n** (number of rows in the beginning having comments to skip)
- While reading file, pass the parameter **skipfooter = n** (number of rows at the end having comments to skip)

In [ ]:
# The following file has three lines of comments in the beginning of the file.
df = pd.read_excel(io='datasets/classmarkswithcomments.xlsx',skiprows=3)

df.head()

### c. Reading Excel Workbook with Multiple Sheets
- By default `pd.read_excel()` function read only the first sheet.
- What if we want to read an Excel file having multiple sheets.
- The `big_mart_sales_with_multiple_sheets.xlsx` is a workbook that contains three sheets for different years data. The sheet names are 1985, 1987, and 1997

In [18]:
df = pd.read_excel('datasets/big_mart_sales_with_multiple_sheets.xlsx')
# if you check/view the data you can see, it only contains the data of first excel sheet (for the year 1985)
df.shape
df


Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0              FDP10          NaN          Low Fat         0.127470   
1              DRI11          NaN          Low Fat         0.034238   
2              FDW12          NaN          Regular         0.035400   
3              FDC37          NaN          Low Fat         0.057557   
4              FDC14          NaN          Regular         0.072222   
...              ...          ...              ...              ...   
1458           DRK37          NaN          Low Fat         0.043792   
1459           DRG13          NaN          Low Fat         0.037006   
1460           NCN14          NaN          Low Fat         0.091473   
1461           FDU44          NaN          Regular         0.102296   
1462           NCN18          NaN          Low Fat         0.124111   

                  Item_Type  Item_MRP Outlet_Identifier  \
0               Snack Foods  107.7622            OUT027   
1               Hard Drinks  113.2834            OUT027   
2              Baking Goods  144.5444            OUT027   
3              Baking Goods  107.6938            OUT019   
4                    Canned   43.6454            OUT019   
...                     ...       ...               ...   
1458            Soft Drinks  189.0530            OUT027   
1459            Soft Drinks  164.7526            OUT027   
1460                 Others  184.6608            OUT027   
1461  Fruits and Vegetables  162.3552            OUT019   
1462              Household  111.7544            OUT027   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                          1985      Medium               Tier 3   
1                          1985      Medium               Tier 3   
2                          1985      Medium               Tier 3   
3                          1985       Small               Tier 1   
4                          1985       Small               Tier 1   
...                         ...         ...                  ...   
1458                       1985      Medium               Tier 3   
1459                       1985      Medium               Tier 3   
1460                       1985      Medium               Tier 3   
1461                       1985       Small               Tier 1   
1462                       1985      Medium               Tier 3   

            Outlet_Type  Item_Outlet_Sales  
0     Supermarket Type3          4022.7636  
1     Supermarket Type3          2303.6680  
2     Supermarket Type3          4064.0432  
3         Grocery Store           214.3876  
4         Grocery Store           125.8362  
...                 ...                ...  
1458  Supermarket Type3          6261.8490  
1459  Supermarket Type3          4111.3150  
1460  Supermarket Type3          2756.4120  
1461      Grocery Store           487.3656  
1462  Supermarket Type3          4138.6128  

[1463 rows x 12 columns]

In [13]:
df_1985 = pd.read_excel('datasets/big_mart_sales_with_multiple_sheets.xlsx',sheet_name='1985')
df_1987 = pd.read_excel('datasets/big_mart_sales_with_multiple_sheets.xlsx',sheet_name='1987')
df_1997 = pd.read_excel('datasets/big_mart_sales_with_multiple_sheets.xlsx',sheet_name='1997')

In [19]:
df_1985.shape

(1463, 12)

In [20]:
df_1987.shape

(932, 12)

In [21]:
df_1997.shape

(930, 12)

## 5. Writing Contents of Dataframe to an EXCEL File
- The `pd.to_excel()` method is used to write the contents of a dataframe (with indices) to an Excel file.
- The only required argument is the file path.
- For details see help page or python documents (link given above)

>- Let us create a new single dataframe after concatenating all the above three dataframes using `pd.concat()` method

In [ ]:
df_concatenated = pd.concat(objs=[df_1985, df_1987, df_1997])

df_concatenated.shape

**Note the total number of rows in this dataframe equals to `1463+932+930 = 3325`**

In [ ]:
df_concatenated.head()

In [ ]:
# you can store the concatenated data inside your dataframe into a single Excel file
# You can mention the argument `index= false` for not storing row indices (0, 1,2,3,... in the Excel file.

df_concatenated.to_excel(excel_writer='temp.xlsx', index=False)

In [ ]:
# Let us verify
data = pd.read_excel(io='temp.xlsx')

data.shape

## 6. I/O with JSON Files

>**JSON**: JavaScript Object Notation is a text-based open standard file format that uses human-readable text consisting of attribute–value pairs and arrays. It is a data interchange format that is used to store and transfer the data via Internet, primarily between a web client and a server.

### a. Reading a Simple JSON File

In [ ]:
import sys
!{sys.executable} -m pip install SQLAlchemy psycopg2-binary

In [ ]:
! cat datasets/simple.json

In [ ]:
# read the json file using read_json method of pandas library
df = pd.read_json('datasets/simple.json')
df

### b. Reading JSON File having each record in a separate line
- Some of the json files are written as records i.e each json line is a separate json object. For example:
```
{ 'name' : 'Ahsan', 'roll_no' : '100' } # line 1
{ 'name' : 'Ayesha' , 'roll_no' : '101' } # line 2
```

In [ ]:
! cat datasets/simple_records.json

In [22]:
# To read such file you need to pass `lines=True` to the `read_json()` method of dataframe
df = pd.read_json('datasets/simple_records.json',lines=True)

df

name  age grade
0   Kamal   12     A
1  Hashim   18     B
2  Salman   11     A
3  Mazhar   12     C
4   Eisha   13     B
5  Farhan   22     C
6  Mohsin   11     A
7   Bilal   19     A
8  Ishaan   10     D
9  Zalaid    9     B

## 7. Writing Contents of Dataframe to a JSON File

In [ ]:
df.to_json('datasets/temp.json')

In [23]:
df = pd.read_json('datasets/temp.json')
df

name  age grade
0   Kamal   12     A
1  Hashim   18     B
2  Salman   11     A
3  Mazhar   12     C
4   Eisha   13     B
5  Farhan   22     C
6  Mohsin   11     A
7   Bilal   19     A
8  Ishaan   10     D
9  Zalaid    9     B